### DISCUSSION TITLE
Understanding valhalla memory consumption profile

### DISCUSSION THREAD

**Discussion by ktnr** [valhalla/valhalla#4816] [General]
I would like to better understand the memory consumption profile of valhalla, especially regarding matrix algorithms. I have written tests to gain insights, which raised some questions.

I know valhalla may not be suited for to compute large, geographically distributed matrix request super fast, but it is surprisingly good considering its runtime flexibility. And with chunking large matrices into smaller ones and parallelizing, even large matrices can be calculated fairly quickly. Although memory consumption becomes an obstacle when parallelizing.

I don't know if this is an issue, so I post it here and you can convert it to an issue if you feel like it.

# What I believe to understand

There are two main components to memory consumption:
1. Accessing the graph/tile data via Baldr/Mjolnir. Two options here, either via the tile dir (`mjolnir.tile_dir`) or via memory mapped tar file (`mjolnir.tile_extract`). Tile cache size is managed by `mjolnir.max_cache_size`.
2. Runtime components of the algorithms such as edge labels via Thor. To a certain extent, this can be managed via config options, e.g., `thor.max_reserved_labels_count_astar` or `thor.clear_reserved_memory`.

Let's assume we have a country extract running with a single thread. When I route on data in a small geographic region, only those tile data are loaded and not that of the whole country. When the requests are spread out over the whole country, more memory will have to be consumed by tile data. So that in the worst case the whole tile directory is loaded into memory (limited by `mjolnir.max_cache_size` with a default value of 1GB).

The runtime components reserve as much memory as they require and are cleared afterwards. As @dnesbitt61 [explained](https://github.com/valhalla/valhalla/issues/3556#issuecomment-1058291344):

> With 512x512 matrix there would be 1024 EdgeLabel lists, 1024 adjacency lists (double bucket queue), 1024 EdgeStatus lookups - I suspect the majority of memory use comes from those temporary structures. The way CostMatrix is called within matrix_action.cc it seems that all temporary memory would be cleaned up when the costmatrix object goes out of scope...

# Expectations

## Single thread

Regarding:
1. Graph data: For example, the `valhalla_tiles.tar` of Germany is about 4.6 GB. The tile tar is memory mapped and the required data are loaded dynamically (?) into memory and the tile cache should not significantly exceed its default limit of 1 GB. 
2. Runtime components: During routing an arbitrary amount of memory may be used, but after routing the used memory should be freed up completely (`thor.clear_reserved_memory`) or up to a certain threshold (`thor.max_reserved_labels_count_astar`).

## Multiple threads

1. Graph data: When using `mjolnir.tile_extract` the tile data is memory mapped and accessible from all threads. The memory usage for tile data should be the same as in the single threaded context.
2. Runtime components: The runtime components need to be allocated for each thread, so they scale with the number of threads.

We have memory consumption `~= tiles + number threads * matrix data structures`.

# Observations

## Setup

- built e5e695f5209440aabea02004fa3999f49ca154de from source and ran `valhalla_service`. 
- the default `valhalla.json` is slightly modified to allow 400x400 matrix request and distances across Germany.
- memory mapped tar is used: `[INFO] Tile extract successfully loaded with tile count: 1205`
- 16 GBs of RAM

Files to reproduce:
- valhalla config [valhalla-germany.json](https://github.com/user-attachments/files/16372685/valhalla-germany.json) running on an extract of Germany
- memory monitoring script including valgrind/massif [monitor_memory_valgrind.sh.txt](https://github.com/user-attachments/files/16372709/monitor_memory_valgrind.sh.txt)
- coordinates in Germany for testing [locations-germany.zip](https://github.com/user-attachments/files/16372801/locations-germany.zip) and script to test [test-routingpy.py.txt](https://github.com/user-attachments/files/16372833/test-routingpy.py.txt)
- memory monitoring results for the single threaded [germany-100.zip](https://github.com/user-attachments/files/16372839/germany-100.zip) and multi-threaded [germany-100-x2-sequential.zip](https://github.com/user-attachments/files/16372843/germany-100-x2-sequential.zip) run

## Single thread

I run the following matrix requests in order with memory usage peak / after (initial memory usage before starting valhalla: 4.4 GB):
- 100 random locations all over Germany: 6.8 GB / 6.0 GB
- The same 100 locations again: 6.8 GB / 6.0 GB
- 200 locations: 9.2 GB / 7.6 GB
- Again the same request with 100 locations: 8.4 GB / 7.6 GB (I would have expected the memory usage to not increase significantly, because the request is smaller than the previous one, so enough space had already been reserved, and idle memory usage before starting the request is above peak level of the initial 100 location request)
- 300 locations: 11.5 GB / 9.0 GB
- The same 100 locations: 9.8 GB / 9.0 GB 
- 400 locations: 13.0 GB / 10 GB

Is it expected that large requests permanently increase memory that will not be released afterwards, so that subsequent smaller request lead to an additional memory increase?

I reran the 100 location request with `valgrind --tool=massif` and other memory logging tools:
- `meminfo` shows around 1.15 GB under "Mapped" for a single and two threads. Should the mapped memory not be around 4.6 GB but the actual cached memory be ~1 GB? This may also be a misunderstanding regarding what `meminfo` reports under "Mapped"...
- while only 2.4 GB of memory is used at peak, `massif` reports about 25 GB of heap allocated/reserved memory where 95% is from cost matrix runtime components.

```
|     ->94.53% (25,600,000,000B) 0x4E9EAC: reserve (vector.tcc:79)
|       ->94.53% (25,600,000,000B) 0x4E9EAC: valhalla::thor::CostMatrix::Initialize(google::protobuf::RepeatedPtrField<valhalla::Location> const&, google::protobuf::RepeatedPtrField<valhalla::Location> const&, valhalla::Matrix const&) (costmatrix.cc:334)
|         ->94.53% (25,600,000,000B) 0x4EE5A1: valhalla::thor::CostMatrix::SourceToTarget(valhalla::Api&, valhalla::baldr::GraphReader&, std::array<std::shared_ptr<valhalla::sif::DynamicCost>, 4ul> const&, valhalla::sif::TravelMode, float) (costmatrix.cc:137)
```

How can we square the peak memory usage of about 2.4 GB with the allocated/reserved memory of 25 GB from `massif`? I'm no expert but read that
- `massif` tracks all memory allocations and reservations made by the program, including virtual memory. These can include allocations that are not fully backed by physical memory at any given time.
- the Linux kernel can overcommit memory, allowing programs to allocate more virtual memory than is physically available.

Should the reserved limits be reduced?

From a different run with more debugging symbols the allocation can be traced back to `sif::BDEdgeLabel`:
```
99.95% (39,308,702,643B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->97.64% (38,400,000,000B) 0x40BA55: std::vector<valhalla::sif::BDEdgeLabel, std::allocator<valhalla::sif::BDEdgeLabel> >::reserve(unsigned long) (in /home/ktnr/source/valhalla/build/valhalla_service)
| ->48.82% (19,200,000,000B) 0x41804D: valhalla::thor::CostMatrix::Initialize(google::protobuf::RepeatedPtrField<valhalla::Location> const&, google::protobuf::RepeatedPtrField<valhalla::Location> const&) (in /home/ktnr/source/valhalla/build/valhalla_service)
```

## Multiple threads

Now I run `valhalla_service valhalla.json 2` with `massif` and compare against the single threaded run.

I run the following matrix requests in order with memory usage peak / after (initial memory usage before starting valhalla 4.4 GB):
- First run of 100 random locations all over Germany: 6.8 GB / 6.0 GB (allocation of the first thread)
- Second run of the same 100 locations: 8.4 GB / 7.6 GB (increase is the additional allocation of the second thread)
- Running the same request in parallel: 9.2 GB / 7.6 GB

After the run of the first thread, we have 6.0 - 4.4 = 1.6 GB in idle usage. After the run of the second thread, idle usage increases by another 1.6 GB. This does not square with `~= tiles + number threads * matrix data structures`. Where do the 1.6 GB per thread in idle usage come from? Temporary peak usage is an additional 0.8 GB per thread.

Valgrind shows 31 GB of reserved memory compared to 25 GB for the single threaded run. Because 95% is memory from runtime components from cost matrix, I would have expected 2 * 25GB = 50 GB: 
```
->95.90% (32,000,000,000B) 0x4E9EAC: valhalla::thor::CostMatrix::Initialize
```

## Surprises

Playing around with `thor.clear_reserved_memory`, `thor.max_reserved_labels_count_astar` etc. didn't fully go as expected. 

I tried the 100 location request:
1. `thor.clear_reserved_memory = false` and default `max_reserved_labels_count_*` (reference): 6.8 GB / 6.0 GB
1. `thor.clear_reserved_memory = true`: 6.8 GB / 5.9 GB
1. `thor.clear_reserved_memory = false` and increased `max_reserved_labels_count_*` by a factor of 10: 6.8 GB / 6.0 GB
1. `thor.clear_reserved_memory = true` and increased `max_reserved_labels_count_*` by a factor of 10: 6.8 GB / 5.9 GB
1. `thor.clear_reserved_memory = false` and decreased `max_reserved_labels_count_*` by a factor of 1000: 7.0 GB / 6.9 GB
1. `thor.clear_reserved_memory = true` and decreased `max_reserved_labels_count_*` by a factor of 1000: 7.0 GB / 6.6 GB

The small differences between 2), 3) and 4) may be measurement errors. However, glancing at the code, I guess I was expecting that
- there would be a bigger difference between 2) and 3)
- 2), 4) and 6) would yield about the same idle memory

For reference, with the limits in the default config of 1M to 4M max reserve labels and taking (EdgeLabel: 48 Byte; BDEdgeLabel: 48 B + 32 B) 80 B per edge label, we should be at 80 MB to 320 MB per label group at runtime.

# Summary

1. Is it expected that large requests permanently increase memory that will not be released afterwards, so that subsequent smaller request lead to an additional memory increase?
1. Why do heap reservations exceed RAM usage by a factor of 10? Should the reserved limits be reduced?
1. How can the observed memory consumption for multiple threads be squared with `~= tiles + number threads * matrix data structures`?
1. Is the observed behavior with `thor.clear_reserved_memory = true` expected?
1. Is there a third component that contributes significantly to memory consumption?

**Comment by kevinkreiser** [comment]
first clarification:

> So that in the worst case the whole tile directory is loaded into memory (limited by mjolnir.max_cache_size with a default value of 1GB).

`max_cache_size` only cares about the memory that isn't memory mapped, that is the memory that is not shared by various threads. so if you are using the `tile_extract` then you will likely be able to load a good chunk of the planet before a gig is reached on each thread. also similar to what dave said about the matrix, we allow going over the max_cache_size to get work done but in ever iteration of the algorithm, if we go over it we will deallocate tiles in the cache to get back under the threshold. in the simple case (not using global synchronized cache, that has an LRU) i think it just purges all the tiles in the cache.

**Comment by kevinkreiser** [comment]
> Is it expected that large requests permanently increase memory that will not be released afterwards, so that subsequent smaller request lead to an additional memory increase?

it is possible and depends on the config parameters you noted that control the edge label buckets in the algorithms. @nilsnolde recently did a bunch of work on that stuff so its possible we increased a non configurable limit that isnt reached by your requests. i would expect though that with large enough requests reaching all threads that you'd hit a max that it wouldnt increase past. at any rate, we can and should make it configurable if its not (in the same way we do for regular a to b routing). id have to go check but its trivial for us to add it if its not fully configurable

**Comment by answerquest** [comment]
Hi, thanks to OP for laying this out. I've had a similar experience:  
Setting `thor.clear_reserved_memory = true` in valhalla.json config has no effect : once a large matrix / route / optimize query is processed and done with, the program never reverts back to its earlier memory footprint. It stays at the maximum. And this is with a smaller region (country or sub-country level pbf). 

This sucks because there's other valhalla instances on the same server in my use case, supposed to share the same RAM, each having their minimums and then a common chunk for burst usage.  

Usage Profile:
- Majority of day: v.low usage.  
- Only for few seconds in the day: Few large requests.  
- And not for all instances at the same time (and we have capability to make them wait in line).  

But now once one of the instances gets their 1 big request, thereafter the entire RAM of the server is blocked off by them indefinitely, preventing other instances from functioning.

It would be great to have some command we could issue to the running program to kind of "reboot" internally, clear everything that was loaded out and have a fresh start.  I wished this `thor.clear_reserved_memory = true` would do that.

If anyone is able to figure out any combination of settings that helps to clear out the burst-time memory occupation, it would be most helpful.

In absence of this, I have to work on restarting the valhalla serve instances externally, which involves a few seconds of downtime.

**Comment by nilsnolde** [comment]
this has been asked/reported in multiple repos and multiple issues each. a few of them have the right answer: play around with `max_reserved_labels_count_bidir_dijkstras` and possibly `max_reserved_locations_costmatrix`.

**Comment by answerquest** [comment]
Thanks; for others looking, here's the issue where this is discussed in more detail: https://github.com/valhalla/valhalla/issues/5172

**Comment by ktnr** [comment]
According to https://github.com/valhalla/valhalla/issues/5172, it seems this is not yet resolved.

Note that I believe I have only investigated `max_reserved_labels_*`, and not with `thor.max_reserved_locations_costmatrix`. Note, the latter has recently been moved to `thor.cost_matrix.max_reserved_locations` per https://github.com/valhalla/valhalla/issues/5172#issuecomment-2764667837 resp. 32cba4fed382efa944e01f6f1eec8c92e3e38166.

**Comment by ddaeschler** [comment]
See https://github.com/valhalla/valhalla/issues/5172#issuecomment-4014353130 For me this is related to the number of workers and can be capped by limiting the max workers when starting valhalla. For my use case memory usage could end up capping out around 2.5GB per worker.
